### PR TITLE
[FEATURE] Allow persisting to composer storage paths

### PR DESCRIPTION
**Pull request by ohader**
Screenshot 2019-11-19 at 01 02 35

* allows to persist new extensions to composer-path directory (retrieved from root composer.json repositories section `path` in case `symlink` is enabled (→ if not explicitly disabled)
* corresponding composer.json example
```
    "repositories": {
        "local": {
            "type": "path",
            "url": "packages/*"
        }
    },
```

### CODE REVIEW

File: Classes/Controller/BuilderModuleController.php [hunk 1]
```diff
@@ -262,10 +278,22 @@ protected function rpcActionSave()
...
         $extensionDirectory = $extension->getExtensionDir();
+        $publicExtensionDirectory = Environment::getExtensionsPath() . '/' . $extension->getExtensionKey();
+        $usesComposerPath = $this->extensionService->isComposerStoragePath($extensionDirectory);
+        $extensionExistedBefore = is_dir($extensionDirectory);
 
-        if (!is_dir($extensionDirectory)) {
+        if (!$extensionExistedBefore) {
             GeneralUtility::mkdir($extensionDirectory);
-        } else {
+            $extensionExistedBefore = false;
```
Comment (liayn) [inline Classes/Controller/BuilderModuleController.php]: Why setting this to `false` again if the condition above is already ensuring this is `false`?

File: Classes/Service/ExtensionService.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                $storagePaths[] = preg_replace('#/[*?/]+$#', '', $repositoryPath);
+            }
+        }
+        return $storagePaths;
+    }
+
+    /**
+     * @param string $path
+     * @return bool
+     */
```
Comment (liayn) [inline Classes/Service/ExtensionService.php]: I would remove such doc headers unless they convey more information than the method signature already does. (DRY principle)

File: Classes/Service/ExtensionService.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    public function isComposerStoragePath(string $path): bool
+    {
+        foreach ($this->resolveComposerStoragePaths() as $composerStoragePath) {
+            if (StringUtility::beginsWith($path, $composerStoragePath)) {
```
Comment (liayn) [inline Classes/Service/ExtensionService.php]: The `beginsWith` method is pretty expensive due to its number of sanity checks. We know we are dealing real, non-empty strings here, so I suggestion to go with `strpos() === 0` directly here

File: Resources/Public/jsDomainModeling/wireit/js/WiringEditor.js [hunk 4]
```diff
@@ -658,25 +659,30 @@
...
 							console.log(ex);
 						}
 					}, this, true);
 					list.appendChild(li);
 				}
-			}
+			} else {
+			  list = document.createElement('div');
+			  list.innerText = 'No extensions found...';
+      }
```
Comment (liayn) [inline Resources/Public/jsDomainModeling/wireit/js/WiringEditor.js]: Indentation is a bit off here

Comment (ohader) [reply Resources/Public/jsDomainModeling/wireit/js/WiringEditor.js]: EditorConfig was different to actual style used in this file...

File: Classes/Service/ExtensionService.php (hunk 3 above)
Comment (ohader) [reply Classes/Service/ExtensionService.php]: Fine, initially I thought about that as well... ;-)

File: Resources/Public/jsDomainModeling/wireit/js/WiringEditor.js (hunk 4 above)
Comment (liayn) [reply Resources/Public/jsDomainModeling/wireit/js/WiringEditor.js]: Good point. I updated the editorconfig some time ago, but we did not adjust all files.

File: Classes/Service/ExtensionService.php (hunk 2 above)
Comment (liayn) [reply Classes/Service/ExtensionService.php]: This comment is actually meant as a general suggestion for all new introduced code. You removed it for this single place now, thanks. I guess I'll do a dedicated cleanup round to get rid of all others and existing docblocks

**Pull request merged**